File: services/python-microservices/cart-service/app/infrastructure/web/dto.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from decimal import Decimal


class AddItemRequest(BaseModel):
    product_id: str
    quantity: int = 1


class ApplyCouponRequest(BaseModel):
    coupon_code: str


class CartItemResponse(BaseModel):
    id: str
    product_id: str
    product_name: str
    price: Decimal
    quantity: int


class CartResponse(BaseModel):
    id: str
    user_id: str
    status: str
    coupon_code: Optional[str] = None
    discount_percent: Decimal
    items: list[CartItemResponse]
    subtotal: Decimal
    total: Decimal
    created_at: datetime
    updated_at: datetime

File: services/nodejs-microservices/cart-service/src/infrastructure/web/dto.ts
import { Cart, CartItem } from '../../domain/models/Cart';
import { CartSummary } from '../../domain/interfaces/ICartService';

export interface CreateCartRequest { userId: string; }
export interface AddItemRequest { productId: string; quantity: number; }
export interface UpdateQuantityRequest { quantity: number; }
export interface ApplyCouponRequest { couponCode: string; }

export function toCartSummaryResponse(summary: CartSummary) {
  return {
    id: summary.cart.id,
    userId: summary.cart.userId,
    status: summary.cart.status,
    couponCode: summary.cart.couponCode,
    discountPercent: summary.cart.discountPercent,
    items: summary.items.map(item => ({
      id: item.id,
      productId: item.productId,
      productName: item.productName,
      quantity: item.quantity,
      unitPrice: item.unitPrice,
      subtotal: item.unitPrice * item.quantity,
    })),
    subtotal: summary.subtotal,
    discount: summary.discount,
    total: summary.total,
  };
}
